### ISSUE TITLE
Bert manually fp16 loss scale

### ISSUE THREAD

**Issue by YangFei1990** [NVIDIA/DeepLearningExamples#139]
I found that the loss was cast back to fp32 [here](https://github.com/NVIDIA/DeepLearningExamples/blob/d9f925cb9c5628578e68ea787365dec5925af061/TensorFlow/LanguageModeling/BERT/run_pretraining.py#L293). Also, when I run the pre-training, I printed out the tensor type of gradients, they are fp32. 
Given all gradients and cost are fp32, why do we still need do loss scale on it?

**Comment by swethmandava** [comment]
The gradients should be in fp16. Are you using manual fp16 or AMP? 

If AMP, it takes care of the graph rewrite for you.
If manually, we cast to fp16 based on the variable here https://github.com/NVIDIA/DeepLearningExamples/blob/d9f925cb9c5628578e68ea787365dec5925af061/TensorFlow/LanguageModeling/BERT/modeling.py#L140

**Comment by YangFei1990** [comment]
I'm using manual fp16. I printed out the tensor type from compute gradients and found they were fp32

**Comment by YangFei1990** [comment]
I watched a nvidia youtube video [here](https://www.youtube.com/watch?v=i1fIBtdhjIg&t=233s), there is a picture:
<img width="708" alt="Screen Shot 2019-07-30 at 5 22 03 PM" src="https://user-images.githubusercontent.com/33940270/62174397-aa432b00-b2ee-11e9-8269-0b4c3f2b81dd.png">
But I can't find the step to cast the loss back to fp16. How do you guarantee that the gradients are fp16?

**Comment by YangFei1990** [comment]
Is there any update on this issue?

**Comment by swethmandava** [comment]
Check this out for more information: https://docs.nvidia.com/deeplearning/frameworks/tensorflow-user-guide/index.html#tfamp-faq
